### DISCUSSION TITLE
01c_forecasting_hierarchical_global.ipynb

### DISCUSSION THREAD

**Discussion by pymcq** [alan-turing-institute/sktime#3420] [Q&A]
Hi,

I am excited to test out sktime as part of a school project, but when I'm working through the "Hierarchical, Global, and Panel Forecasting with sktime" notebook with a relatively simple starter example, I'm running into an Assertion Error (it's rather long, so I added to the bottom of this message).

Before asking for help, I tried for due diligence...another new user reported a similar issue related to the frequency of the DateTimeIndex (https://github.com/alan-turing-institute/sktime/issues/1212), but I can't figure out how to set the frequency for a DatetimeIndex when it's part of a MultiIndex to save my life!  Also the error message is somewhat different, so I'm not sure that's even the issue. I feel so close,  yet so far....so any help anyone could offer would be much appreciated!

Note multiple stores with multiple product categories are involved so dates are repeated.
​I'm extracting the data from a simple .csv file, and then changing the data type to date:
`train_data["date"] = pd.to_datetime(train_data["date"],errors='coerce')`
 
I set up my multi-index:
`train_h.set_index(["store_nbr", "family", "date"], inplace=True)`

Double checked my time point landed where expected:
`train_h.index.get_level_values(level=-1)`

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06', '2013-01-07', '2013-01-08',
               '2013-01-09', '2013-01-10',
               ...
               '2017-08-06', '2017-08-07', '2017-08-08', '2017-08-09',
               '2017-08-10', '2017-08-11', '2017-08-12', '2017-08-13',
               '2017-08-14', '2017-08-15'],
              dtype='datetime64[ns]', name='date', length=277860, freq=None)

Set my forecasting horizon:
`horizon = ForecastingHorizon(pd.PeriodIndex(pd.date_range("2017-08-16", periods=16, freq="D")), is_relative=False)
horizon`
ForecastingHorizon(['2017-08-16', '2017-08-17', '2017-08-18', '2017-08-19',
             '2017-08-20', '2017-08-21', '2017-08-22', '2017-08-23',
             '2017-08-24', '2017-08-25', '2017-08-26', '2017-08-27',
             '2017-08-28', '2017-08-29', '2017-08-30', '2017-08-31'],
            dtype='period[D]', is_relative=False)

Selected a very simple forecaster to start...
`forecaster = NaiveForecaster(strategy="last")`

And then tried to run the forecast
`next15 = forecaster.fit(train_h, fh=horizon).predict()`

and boom!

`---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-17-4a9d3041eaef> in <module>
----> 1 next15 = forecaster.fit(train_h, fh=horizon).predict()

~\anaconda3\lib\site-packages\sktime\forecasting\base\_base.py in predict(self, fh, X)
    357         else:
    358             # otherwise we call the vectorized version of predict
--> 359             y_pred = self._vectorize("predict", X=X_inner, fh=fh)
    360 
    361         # convert to output mtype, identical with last y mtype seen

~\anaconda3\lib\site-packages\sktime\forecasting\base\_base.py in _vectorize(self, methodname, **kwargs)
   1660 
   1661                 if methodname != "update_predict_single":
-> 1662                     y_preds += [method(X=Xs[i], **kwargs)]
   1663                 else:
   1664                     y_preds += [method(y=ys[ix], X=Xs[i], **kwargs)]

~\anaconda3\lib\site-packages\sktime\forecasting\base\_base.py in predict(self, fh, X)
    354         # we call the ordinary _predict if no looping/vectorization needed
    355         if not self._is_vectorized:
--> 356             y_pred = self._predict(fh=fh, X=X_inner)
    357         else:
    358             # otherwise we call the vectorized version of predict

~\anaconda3\lib\site-packages\sktime\forecasting\naive.py in _predict(self, fh, X)
    314             Exogenous time series
    315         """
--> 316         y_pred = super(NaiveForecaster, self)._predict(fh=fh, X=X)
    317 
    318         # test_predict_time_index_in_sample_full[ForecastingPipeline-0-int-int-True]

~\anaconda3\lib\site-packages\sktime\forecasting\base\_sktime.py in _predict(self, fh, X)
     28 
     29         # all values are out-of-sample
---> 30         if fh.is_all_out_of_sample(self.cutoff):
     31             return self._predict_fixed_cutoff(
     32                 fh.to_out_of_sample(self.cutoff), **kwargs

~\anaconda3\lib\site-packages\sktime\forecasting\base\_fh.py in is_all_out_of_sample(self, cutoff)
    611             cutoff.
    612         """
--> 613         return sum(self._is_out_of_sample(cutoff)) == len(self)
    614 
    615     def to_indexer(self, cutoff=None, from_cutoff=True):

~\anaconda3\lib\site-packages\sktime\forecasting\base\_fh.py in _is_out_of_sample(self, cutoff)
    594     def _is_out_of_sample(self, cutoff=None) -> np.ndarray:
    595         """Get index location of out-of-sample values."""
--> 596         return np.logical_not(self._is_in_sample(cutoff))
    597 
    598     def is_all_out_of_sample(self, cutoff=None) -> bool:

~\anaconda3\lib\site-packages\sktime\forecasting\base\_fh.py in _is_in_sample(self, cutoff)
    573     def _is_in_sample(self, cutoff=None) -> np.ndarray:
    574         """Get index location of in-sample values."""
--> 575         relative = self.to_relative(cutoff).to_pandas()
    576         null = 0 if is_integer_index(relative) else pd.Timedelta(0)
    577         return relative <= null

~\anaconda3\lib\site-packages\sktime\forecasting\base\_fh.py in to_relative(self, cutoff)
    460         """
    461         cutoff = self._coerce_cutoff_to_index_element(cutoff)
--> 462         return _to_relative(fh=self, cutoff=cutoff)
    463 
    464     def to_absolute(self, cutoff):

~\anaconda3\lib\site-packages\sktime\forecasting\base\_fh.py in _to_relative(fh, cutoff)
    681     else:
    682         absolute = fh.to_pandas()
--> 683         _check_cutoff(cutoff, absolute)
    684 
    685         if isinstance(absolute, pd.DatetimeIndex):

~\anaconda3\lib\site-packages\sktime\forecasting\base\_fh.py in _check_cutoff(cutoff, index)
    775 
    776     if isinstance(index, pd.PeriodIndex):
--> 777         assert isinstance(cutoff, pd.Period)
    778         assert index.freqstr == cutoff.freqstr
    779 

AssertionError: 

​`

**Comment by fkiraly** [comment]
hm, might be a bug?

Can you kindly post an entire block of code that reliably reproduces the issue?

You can post it as a python block in markdown, for that use three apostrophes (like in your `in-line blocks` but three at start/end) and the first three apostrophes followed by the string `python` (no space between apostrophes and it, then newline, past the code between the two lines with the three apostrophes).

**Reply by fkiraly** [reply]
example:
```python
print("hello world")
print("more code")
```

**Comment by pymcq** [comment]
Thank you!  The block below reliably produces the issue. 

Note this is my most recent slice & dice to pass a clean set in case that would (understandably) be an issue.
(all stores & families have the same number of expected rows so no missing dates...)
`train_sub = train_data.loc[(train_data.store_nbr <= 5) & (train_data.date > "2017-01-01")]`

```python
#import pandas library for data analysis & manipulation
import pandas as pd

#import pandas library for scientific computing
import numpy as np

#training data for model
train_path = "***\\Project_Data\\train.csv.zip"
#test data for prediction model
test_path = "***\\Project_Data\\test.csv"

#import data into pandas dataframe
train_data = pd.read_csv(train_path,compression="zip")
test_data = pd.read_csv(test_path)
                            
#check number of rows & columns in each set
print("training set",train_data.shape)
print("testing set", test_data.shape)

pip install sktime

from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.naive import NaiveForecaster

#set date as date type
train_data["date"] = pd.to_datetime(train_data["date"],errors='coerce')
train_data.info()
train_data.head()

train_data = train_data.sort_values(by=["store_nbr", "family", "date"])

train_sub = train_data.loc[(train_data.store_nbr <= 5) & (train_data.date > "2017-01-01")]
train_sub

train_h = train_sub.drop(["id","onpromotion"], axis=1)
train_h

train_h.set_index(["store_nbr", "family", "date"], inplace=True)
train_h

train_h.index.get_level_values(level=-1)

horizon = ForecastingHorizon(pd.PeriodIndex(pd.date_range("2017-08-16", periods=16, freq="D")), is_relative=False)
horizon

forecaster = NaiveForecaster(strategy="last")
next15 = forecaster.fit(train_h, fh=horizon).predict()
```

**Reply by fkiraly** [reply]
can you provide this without:

* any data loaded externally, i.e., `read_csv`
* any changes to the environment, i.e., `pip install sktime`

(so we can run it without loading data or changing the environment)

thanks

**Reply by pymcq** [reply]
Thank you for the quick response! 
I took another run at the problem & the issue was the missing frequency in the training data.
Now it runs beautifully!  👍 

Revised code below for anyone else who runs into this issue... I left the read_csv in but commented out (changes are relevant to solution).   The original error message was quite cryptic & intimidating...if possible to add clarification, likely would help future users. 
I appreciate the time you have taken to respond, now I will go play around more with sktime!

```python
#import pandas library for data analysis & manipulation
import pandas as pd

#import pandas library for scientific computing
import numpy as np

#import data into pandas dataframe
#train_data = pd.read_csv(train_path, 
                        # usecols=["store_nbr", "family", "date", "sales"],
                        # dtype={"store_nbr":"int",
                               # "family":"object",
                               # "sales":"float32",},
                                #parse_dates=["date"],
                               #infer_datetime_format=True,
                                #compression="zip")

train_data["date"] = train_data.date.dt.to_period("D")
train_data

from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.naive import NaiveForecaster

train_data.info()

train_data = train_data.sort_values(by=["store_nbr", "family", "date"])

train_sub = train_data.loc[(train_data.store_nbr <= 5) & (train_data.date > "2017-01-01")]
train_sub

train_sub.set_index(["store_nbr", "family", "date"], inplace=True)

train_sub.info()

train_sub.index.get_level_values(level=-1)

horizon = ForecastingHorizon(pd.PeriodIndex(pd.date_range("2017-08-16", periods=16, freq="D")), is_relative=False)
horizon

forecaster = NaiveForecaster(strategy="last")
next15 = forecaster.fit(train_sub, fh=horizon).predict()
next15

```

**Reply by fkiraly** [reply]
Great to hear that it works now!

To make the error message clearer, could you kindly elaborate what exactly the failure condition was, and provide a piece of *failing* code?
So we know where the error message needs to be improved.

**Reply by pymcq** [reply]
Absolutely!  

The error message was raised when I ran the fit/predict line.
`next15 = forecaster.fit(train_sub, fh=horizon).predict()`

..but the data prep in the beginning was the problem.  Essentially, the frequency of the training data needs to match the frequency of the forecast horizon (which makes sense!). Perhaps with further clarification that the frequency needs to be explicitly declared for the training data, a note to this effect at the end of the long message that pops up would be very helpful.

Making this requirement more explicit in the instructions could also help us newcomers out.  I read through the sample notebooks several times to try and figure out what I was doing wrong but I don't recall see this condition mentioned.  I found a clue by searching the issues for "Assertion Errors" and guessing that someone else who posted about the sample notebooks was as clueless as me!

Here's the original issue that pointed me towards a solution:
https://github.com/alan-turing-institute/sktime/issues/1212#issue-952950029

Thanks again for your response!

**Reply by fkiraly** [reply]
could you kindly give an example when it fails? That would be hugely useful in improving the error messaging.
Your original code, or an intentionally broken variant of your "good example" would be helpful.

**Comment by pymcq** [comment]
Sorry...I thought I had & I don't know what else to say.  It wasn't likely a problem a more experienced person would have had. Just consider this a classic "user error" and close.  Thanks
